pkgs/os-specific/linux/nvidia-drivers/default.nix

```diff
@@ -0,0 +1,229 @@
+in
+
+assert any (n: n == buildConfig) [
+  "kernelspace"
+  "userspace"
+  "all"
+];
+assert any (n: n == channel) [
+  "legacy340"
+  "long-lived"
+  "short-lived"
+  "testing"
+];
+assert buildKernelspace -> kernel != null;
+assert nvidiasettingsSupport -> (
+  atk != null &&
+  gdk_pixbuf != null &&
+  glib != null &&
+  pango != null && 
+  (gtk3Support || gtk2Support)
+);
+assert gtk2Support -> (
+  gtk2 != null &&
+  !gtk3Support &&
+  cairo == null &&
+  gtk3 == null
+);
+assert gtk3Support  -> (
+  versionAtLeast versionMajor "346" &&
+  cairo != null &&
+  gtk3 != null &&
+  !gtk2Support &&
+  gtk2 == null
+);
+assert libsOnly -> !buildKernelspace;
+
+stdenv.mkDerivation {
+
+  name = "nvidia-drivers-${buildConfig}-${version}"
+       + "${optionalString buildKernelspace "-${kernel.version}"}";
+
+  src = fetchurl {
+    url =
+      if system == "i686-linux" then
+        "http://us.download.nvidia.com/XFree86/Linux-x86/${version}/" +
+        "NVIDIA-Linux-x86-${version}.run"
+      else if system == "x86_64-linux" then
+        "http://us.download.nvidia.com/XFree86/Linux-x86_64/${version}/" +
+        "NVIDIA-Linux-x86_64-${version}-no-compat32.run"
+      else
+        throw "The NVIDIA drivers do not support the `${system}' platform";
+    sha256 =
+      if system == "i686-linux" then
+        sha256i686
+      else if system == "x86_64-linux" then
+        sha256x86_64
+      else
+        throw "The NVIDIA drivers do not support the `${system}' platform";
+  };
+
+  kernel =
+    if buildKernelspace then
+      kernel.dev
+    else
+      null;
+
+  disallowedReferences =
+    if buildKernelspace then
+      [ kernel.dev ]
+    else
+      [ ];
+
+  patches =
+    optionals (versionMajor == "340") [
+      #./nvidia-340.76-kernel-4.0.patch
+    ] ++
+    optionals (versionAtLeast versionMajor "346" &&
+               versionOlder versionMajor "355") [
+      ./nvidia-4.2.patch
+    ];
+
+  # Make sure anything that isn't declared within the derivation
+  # is inherited so that it is passed to the builder.
+  inherit
+    buildKernelspace
+    buildUserspace
+    libsOnly
+    gtk3Support
+    nvidiasettingsSupport
+    version
+    versionMajor;
+
+  nativeBuildInputs = [ nukeReferences ];
+
+  builder = ./builder.sh;
+
+  dontStrip = true;
+
+  glPath = makeLibraryPath [
+    xlibs.libXext
+    xlibs.libX11
+    xlibs.libXrandr
+  ];
+  programPath = makeLibraryPath [
+    xlibs.libXv
+  ];
+  allLibPath = makeLibraryPath [
+    stdenv.cc.cc
+    xlibs.libX11
+    xlibs.libXext
+    xlibs.libXrandr
+    zlib
+  ];
+  gtkPath = optionalString (!libsOnly) (
+    makeLibraryPath (
+      [
+        atk
+        pango
+        glib
+        gdk_pixbuf
+      ] ++ (
+        if gtk3Support then [
+          cairo
+          gtk3
+        ] else [
+          gtk2
+        ]
+      )
+    )
+  );
+
+  passthru = {
+    inherit
+      version
+      versionMajor;
+    nvenc = (
```

Comment:
Nitpick: I would personally remove parentheses around these two `if`s in `passthru`. They shouldn't be needed and even visually it's expressed by indentation.